benchmarking/tpch/answers.py

```diff
@@ -333,3 +333,354 @@ def decrease(x, y):
+    supplier = get_df("supplier")
+    nation = get_df("nation")
+    part = get_df("part")
+    partsupp = get_df("partsupp")
+    lineitem = get_df("lineitem")
+
+    res_1 = (
+        lineitem.where(
+            (col("L_SHIPDATE") >= datetime.date(1994, 1, 1)) & (col("L_SHIPDATE") < datetime.date(1995, 1, 1))
+        )
+        .groupby("L_PARTKEY", "L_SUPPKEY")
+        .agg(((col("L_QUANTITY") * 0.5).sum()).alias("sum_quantity"))
+    )
+
+    res_2 = nation.where(col("N_NAME") == "CANADA")
+    res_3 = supplier.join(res_2, left_on="S_NATIONKEY", right_on="N_NATIONKEY")
+
+    daft_df = (
+        part.where(col("P_NAME").str.startswith("forest"))
+        .select("P_PARTKEY")
+        .distinct()
+        .join(partsupp, left_on="P_PARTKEY", right_on="PS_PARTKEY")
+        .join(
+            res_1,
+            left_on=["PS_SUPPKEY", "P_PARTKEY"],
+            right_on=["L_SUPPKEY", "L_PARTKEY"],
+        )
+        .where(col("PS_AVAILQTY") > col("sum_quantity"))
+        .select("PS_SUPPKEY")
+        .distinct()
+        .join(res_3, left_on="PS_SUPPKEY", right_on="S_SUPPKEY")
+        .select("S_NAME", "S_ADDRESS")
+        .sort("S_NAME")
+    )
+
+    return daft_df
+
+
+def q21(get_df: GetDFFunc) -> DataFrame:
+    supplier = get_df("supplier")
+    nation = get_df("nation")
+    lineitem = get_df("lineitem")
+    orders = get_df("orders")
+
+    res_1 = (
+        lineitem.select("L_SUPPKEY", "L_ORDERKEY")
+        .distinct()
+        .groupby("L_ORDERKEY")
+        .agg(col("L_SUPPKEY").count().alias("nunique_col"))
+        .where(col("nunique_col") > 1)
+        .join(lineitem.where(col("L_RECEIPTDATE") > col("L_COMMITDATE")), on="L_ORDERKEY")
+    )
+
+    daft_df = (
+        res_1.select("L_SUPPKEY", "L_ORDERKEY")
+        .distinct()
+        .groupby("L_ORDERKEY")
+        .agg(col("L_SUPPKEY").count().alias("nunique_col"))
+        .join(res_1, on="L_ORDERKEY")
+        .join(supplier, left_on="L_SUPPKEY", right_on="S_SUPPKEY")
+        .join(nation, left_on="S_NATIONKEY", right_on="N_NATIONKEY")
+        .join(orders, left_on="L_ORDERKEY", right_on="O_ORDERKEY")
+        .where((col("nunique_col") == 1) & (col("N_NAME") == "SAUDI ARABIA") & (col("O_ORDERSTATUS") == "F"))
+        .groupby("S_NAME")
+        .agg(col("O_ORDERKEY").count().alias("numwait"))
+        .sort(["numwait", "S_NAME"], desc=[True, False])
+        .limit(100)
+    )
+
+    return daft_df
+
+
+def q22(get_df: GetDFFunc) -> DataFrame:
+    orders = get_df("orders")
+    customer = get_df("customer")
+
+    res_1 = (
+        customer.with_column("cntrycode", col("C_PHONE").str.left(2))
+        .where(col("cntrycode").is_in(["13", "31", "23", "29", "30", "18", "17"]))
+        .select("C_ACCTBAL", "C_CUSTKEY", "cntrycode")
+    )
+
+    res_2 = (
+        res_1.where(col("C_ACCTBAL") > 0).agg(col("C_ACCTBAL").mean().alias("avg_acctbal")).with_column("lit", lit(1))
+    )
+
+    res_3 = orders.select("O_CUSTKEY").distinct()
```

Comment:
discussed offline but we can likely remove these `distinct` since we are doing a `not exists` and dont care about the dup values.